enhancements/network/on-prem-configurable-lb-dns-stack.md

```diff
@@ -0,0 +1,194 @@
+
+The [Keepalived and infra DNS components](https://github.com/openshift/machine-config-operator/tree/master/manifests/baremetal) on the bootstrap node will keep running through the MCO.
+However, the services on masters and workers nodes will be managed by a new operator instead of static pods.
+
+The following functionalities are expected as part of the new operator:
+* Introduce a CRD used by the operator to manage the self-hosted stack.
+* Fields in CR spec that control the self-hosted stack configuration.
+* Fields in CR status that expose self-hosted stack status (e.g: which node holds the API-VIP).
+* Run the self-hosted stack components (Kepalived, Coredns-MDNS, HAProxy, MDNS-publisher) as a K8S applications.
+
+## Risks and Mitigations
+
+- Impact on on-prem platforms
+- Upgrades/Downgrades
+
+## Design Details
+
+### Open Questions [optional]
+
+- Is `cluster-hosted-net-services-operator` an acceptable name? 
+
+### cluster-hosted-net-services operator
+The `cluster-hosted-net-services-operator` is a new operator whose operand is a controller for the cluster hosted DNS, LB and VIP services.
+
+### Standard SLO Behavior
+
+The CHNSO is expected to follow the standard behavior of SLO, additionally, since the CHNSO is applicable only to clusters running on on-prem ( 'baremetal', 'openstack', 'vsphere' or 'ovirt') infrastructure,
+It should support also the expected behaviors for "not in use" SLOs similar to the [CBO operator](https://github.com/openshift/enhancements/blob/master/enhancements/baremetal/an-slo-for-baremetal.md#standard-slo-behaviors).
+
+### cluster-hosted-net-services-operator details
+
+The CHNSO should:
+
+- Be a new `openshift/cluster-hosted-net-services-operator` project.
+- Publish an image called `cluster-hosted-net-services-operator`.
+- Add a new `ClusterOperator` with an additional
+  `Disabled` status for non on-prem platforms.
+- Implement a controller reconciling a singleton instance representing the desired cluster hosted LB/DNS/VIP configuration.
+- Do nothing except set `Disabled=true`, `Available=true`, and
+  `Progressing=False` when the `Infrastructure` resource is a platform
+  type other than openstack, baremetal, ovirt or vsphere.
+
+#### Kubelet and CHNSO circular dependency
+
+In order to run the CHNSO components, Kubelet in each node should first join the cluster by communicating with the control plane, however the components provided by the CHNSO are responsible for enabling that.
+As a result, the below circular dependency should be addressed:
+* Kubelet can't talk to the control plane until the CHNSO has started.
+* CHNSO can't start until Kubelet can talk to the control plane.
+
+To break the circular dependency, the CHNSO should start running before bootstrap is completed in order to assure that bootstrap owns the API-VIP. In addition, a new /etc/hosts entry should be added to the nodes using MCO ignition.  This /etc/hosts entry is expected to resolve api-int hostname to the API-VIP address.
```

Comment:
Well, AFAIK. to break the circular dependency the api-int should be resolved to IP address and this IP address should point to K8S control plane, so I assume that VIP being owned by the bootstrap is also relevant here. 
Let me know, if you still think I should update this section

Reply:
I guess maybe I'm splitting hairs here, but to me the dependency isn't circular for the API VIP. CHNSO has no dependency on the API VIP moving to the cluster. It needs to be happy using the bootstrap API. The dependency chain is linear: API comes up on bootstrap and takes API VIP -> CHNSO starts -> Keepalived starts and once the cluster API also starts it can take the API VIP.

The thing that's circular about the api-int record is that for CHNSO to run on a node, that node needs to have registered with the cluster. But if api-int lives in CHNSO, then the node can't register with the cluster until CHNSO runs.

In other words, there's no sequence of operations where the api-int record can live in CHNSO. There _is_ a sequence of operations where the API VIP can be managed by CHNSO, at least after bootstrap. That's the difference to me. One requires breaking a dependency, the other just adds a dependency to the bootstrap process (CHNSO must run before the bootstrap API stops).

I think both of us understand how this needs to work, but I'm concerned that combining the two problems into one section might be confusing to reviewers who haven't been involved in our design discussions.